lib/improver/utilities/mathematical_operations.py

```diff
@@ -0,0 +1,299 @@
+            integrated_cube.coord(self.coord_name_to_integrate).bounds = (
+                zip(lower_bounds, upper_bounds))
+        elif self.direction_of_integration == "negative":
+            integrated_cube = lower_bounds_cube.copy()
+            integrated_cube.coord(self.coord_name_to_integrate).bounds = (
+                zip(lower_bounds, upper_bounds))
+
+        integrated_cube.data = np.zeros(lower_bounds_cube.shape)
+        return upper_bounds_cube, lower_bounds_cube, integrated_cube
+
+    def perform_integration(
+            self, upper_bounds_cube, lower_bounds_cube, integrated_cube):
+        """Perform the integration.
+
+        Integration is performed by firstly defining the stride as the
+        difference between the upper and lower bound. The contribution from
+        the uppermost half of the stride is calculated by multiplying the
+        upper bound value by 0.5 * stride, and the contribution
+        from the lowermost half of the stride is calculated by multiplying the
+        lower bound value by 0.5 * stride. The contribution from the
+        uppermost half of the stride and the bottom half of the stride is
+        summed.
+
+        As the coordinate is progressively integrated, the contribution of
+        each stride is cumulatively summed.
+
+        Args:
+            upper_bounds_cube (iris.cube.Cube):
+                Cube containing the upper bounds to be used during the
+                integration.
+            lower_bounds_cube (iris.cube.Cube):
+                Cube containing the lower bounds to be used during the
+                integration.
+            integrated_cube (iris.cube.Cube):
+                Cube that will be used for storing the output of the
+                integration containing the most appropriate coordinates.
+
+        Returns:
+            integrated_cube (iris.cube.Cube):
+                Cube containing the output from the integration.
+
+        """
+        # Create a zip for looping over.
+        levels_tuple = zip(
+            upper_bounds_cube.slices_over(self.coord_name_to_integrate),
+            lower_bounds_cube.slices_over(self.coord_name_to_integrate),
+            integrated_cube.slices_over(self.coord_name_to_integrate))
+
+        # Perform the integration
+        stride_sum = 0
+        integrated_cubelist = iris.cube.CubeList([])
+        for (upper_bounds_slice, lower_bounds_slice,
+             integrated_slice) in levels_tuple:
+            upper_bound = (
+                upper_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            lower_bound = (
+                lower_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            if not self.start_point and not self.end_point:
+                pass
+            elif self.start_point:
+                if self.direction_of_integration == "positive":
+                    if lower_bound <= self.start_point:
```

Comment:
Should these be < rather than <= . Can you add a few edge case to the unit tests e.g. start_point =5.0 to ensure that this is doing what you think it should be doing.

Reply:
Yes, I've changed these to be either < or > and added a couple of unit tests to try to show this.